### PR TITLE
[SPARK-40895][BUILD] Upgrade arrow to 10.0.0

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
This PR aims to upgrade Apache Arrow to 10.0.0

### Why are the changes needed?
This version bring some bug fix and improvements, the official  release notes as follows:

- https://arrow.apache.org/release/10.0.0.html

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Pass GitHub Actions

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -319,6 +319,17 @@
     </extraJavaTestArgs>
   </properties>
   <repositories>
+    <!-- ASF Staging for testing -->
```
Comment (LuciferYang) [inline pom.xml:322]: for test, will revert later

File: project/SparkBuild.scala [hunk 2]
```diff
@@ -291,6 +291,7 @@ object SparkBuild extends PomBuild {
       // Google Mirror of Maven Central, placed first so that it's used instead of flaky Maven Central.
       // See https://storage-download.googleapis.com/maven-central/index.html for more info.
       "gcs-maven-central-mirror" at "https://maven-central.storage-download.googleapis.com/maven2/",
+      "asf staging" at "https://repository.apache.org/content/repositories/staging/",
```
Comment (LuciferYang) [inline project/SparkBuild.scala:294]: for test, will revert later

**Comment by dongjoon-hyun** [comment]
cc @sunchao

**Review by itholic — COMMENTED** [review]
Upgrading arrow looks good when the test passed

**Comment by bjornjorgensen** [comment]
Is this released? I only found https://github.com/apache/arrow/tree/release-10.0.0-rc0 and their own tests don't run.

**Comment by LuciferYang** [comment]
@bjornjorgensen has been released and is waiting for the official release notes

- https://github.com/apache/arrow/releases/tag/apache-arrow-10.0.0

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you!
Merged to master for Apache Spark 3.4.

**Comment by LuciferYang** [comment]
Thanks @dongjoon-hyun @itholic @bjornjorgensen

**Pull request closed without merging**